### PR TITLE
test: Add test cases related to each mode of SM4

### PR DESCRIPTION
**Pull request by uudiin**
Added test cases for SM4 algorithm CBC, OFB, CFB and CTR mode.
Test data comes from:
https://tools.ietf.org/id/draft-ribose-cfrg-sm4-10.html

Signed-off-by: Tianjia Zhang <tianjia.zhang@linux.alibaba.com>

##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: test/sm4_internal_test.c [hunk 1]
```diff
@@ -75,12 +75,164 @@ static int test_sm4_ecb(void)
...
+            || !TEST_ptr(ctx = EVP_CIPHER_CTX_new())
+            || !TEST_true(EVP_EncryptInit_ex(ctx, cipher, NULL, key, iv))
+            || !TEST_true(EVP_EncryptUpdate(ctx, ctext, &outlen, input, len))
+            || !TEST_true(EVP_EncryptFinal_ex(ctx, ctext + outlen, &tmplen))
+            || !TEST_mem_eq(ctext, len, expected, len))
+        goto err;
+
+    outlen += tmplen;
+    if (!TEST_true(EVP_DecryptInit_ex(ctx, cipher, NULL, key, iv))
+            || !TEST_true(EVP_DecryptUpdate(ctx, ptext, &outlen, ctext, outlen))
```
Comment (InfoHunter) [inline test/sm4_internal_test.c:102]: Why don't use `len` instead of `outlen` directly?

File: test/sm4_internal_test.c [hunk 2]
```diff
@@ -75,12 +75,164 @@ static int test_sm4_ecb(void)
...
+
+static int test_sm4_cipher(const EVP_CIPHER *cipher,
+                           const unsigned char *key,
+                           const unsigned char *iv,
+                           const unsigned char *input,
+                           const unsigned char *expected,
+                           unsigned int len)
+{
+    EVP_CIPHER_CTX *ctx = NULL;
+    unsigned char ctext[1024];
```
Comment (InfoHunter) [inline test/sm4_internal_test.c:87]: Nitpicking: `1024` seems too much adequate for those test cases...

**Comment by paulidale** [comment]
I think that these cases should be added to `test/recipes/30-test_evp_data/evpciph_sm4.txt` not implemented in the internal test.

File: test/sm4_internal_test.c (hunk 2 above)
Comment (uudiin) [reply test/sm4_internal_test.c:87]: That's right, only 80 bytes are used at most.

File: test/sm4_internal_test.c (hunk 1 above)
Comment (uudiin) [reply test/sm4_internal_test.c:102]: From the test results, the outlen and len of CBC mode are not equal, and the direct use of len will also cause the decryption to fail. It seems that the lengths of the ciphertext and plaintext in CBC mode are not equal.

**Comment by uudiin** [comment]
> I think that these cases should be added to `test/recipes/30-test_evp_data/evpciph_sm4.txt` not implemented in the internal test.

I agree, I'm sorry I didn't notice the test data here, this PR should be negligible.

**Pull request closed without merging**